plugins/wallabag.koplugin/main.lua

```diff
@@ -113,6 +114,19 @@ function Wallabag:init()
         if res then self.is_dateparser_available = true end
         self.is_dateparser_checked = true
     end
+
+    ReaderLink:addToExternalLinkDialog("03_wallabag", function(this, link_url)
```

Comment:
I don't understand this. I revised it, as suggested, but get this error when launching: ` Failed to initialize wallabag plugin: plugins/wallabag.koplugin/main.lua:118: attempt to index field 'link' (a nil value) `

Reply:
OK, this plugin is enabled also when not in Reader (ie starting with file manager?). So:
```lua
if self.ui and self.ui.link then
    self.ui.link:addToExternalLinkDialog(...) 
end 
```
?

Reply:
Well I don't understand this, but it works, so that's good enough!

Reply:
If you're in Filemanager, the wallbag plugin is loaded. self.ui is the apps/filemanager.lua instance, and it does not have an attribute named `link`.
If you're in Reader, the wallbag plugin is loaded. self.ui is the apps/reader.lua instance, and it does have an attribute named `link`: the ReaderLink instance.

Reply:
Much like widgets, there's only *one* plugin class, the one initially loaded by `require` (we'll gloss over how pluginloader does this for plugins, as it does not actually use require, because the end result is the same).

On the other hand, much like widgets, you *will* have multiple *instances* of plugins: every time ReaderUI or FileManager starts, *every* active plugins gets a new instance (i.e., `init` gets called). Conversely, when one of them is torn down, the plugins follow suit (much like widgets, they get a `CloseWidget` event).

Like ReaderUI/FileManager modules (whose lifecycle is similar), plugins inherit a pointer to their parent via self.ui. That will either point to the current ReaderUI or FileManager instance, so, if you're using fields/methods specific to only one of them, you need to be careful, as @poire-z explained ;).